src/workerd/io/worker-fs.c++

```diff
@@ -0,0 +1,245 @@
+  const Stat stat() const override {
+    return getDirectory().stat();
+  }
+
+  size_t count(jsg::Lock& js, kj::Maybe<Type> typeFilter = kj::none) const override {
+    return getDirectory().count(js, typeFilter);
+  }
+
+  void forEach(jsg::Lock& js,
+      kj::FunctionParam<bool(jsg::Lock& js, kj::StringPtr, Stat stat)> fn,
+      kj::Maybe<Type> typeFilter = kj::none) const override {
+    getDirectory().forEach(js, fn, typeFilter);
+  }
+
+  kj::Maybe<kj::OneOf<kj::Own<jsg::fs::File>, kj::Own<jsg::fs::Directory>>> tryOpen(
+      jsg::Lock& js, kj::PathPtr path, kj::Maybe<Type> createAs = kj::none) override {
+    return getDirectory().tryOpen(js, path, createAs);
+  }
+
+  kj::Own<jsg::fs::File> transfer(
+      jsg::Lock& js, kj::StringPtr name, kj::Own<jsg::fs::File>&& file) override {
+    return getDirectory().transfer(js, name, kj::mv(file));
+  }
+
+  kj::Own<Directory> transfer(
+      jsg::Lock& js, kj::StringPtr name, kj::Own<jsg::fs::Directory>&& dir) override {
+    return getDirectory().transfer(js, name, kj::mv(dir));
+  }
+
+  bool remove(
+      jsg::Lock& js, kj::PathPtr path, kj::Maybe<RemoveOptions> options = kj::none) override {
+    return getDirectory().remove(js, path, options);
+  }
+
+  kj::StringPtr jsgGetMemoryName() const override {
+    return "LazyDirectory"_kj;
+  }
+
+  size_t jsgGetMemorySelfSize() const override {
+    return sizeof(LazyDirectory);
+  }
+
+  void jsgGetMemoryInfo(jsg::MemoryTracker& tracker) const override {
+    // We will only track the contents of this directory if it has
+    // be lazily loaded.
+    KJ_SWITCH_ONEOF(lazyDir) {
+      KJ_CASE_ONEOF(dir, kj::Own<jsg::fs::Directory>) {
+        dir->jsgGetMemoryInfo(tracker);
+        return;
+      }
+      KJ_CASE_ONEOF(func, kj::Function<kj::Own<jsg::fs::Directory>()>) {
+        // We don't know the contents of this directory  If it has not, then the contents are yet.
+        // Do not track.
+        return;
+      }
+    }
+  }
+
+ private:
+  mutable kj::OneOf<kj::Own<jsg::fs::Directory>, kj::Function<kj::Own<jsg::fs::Directory>()>>
+      lazyDir;
+
+  jsg::fs::Directory& getDirectory() const {
+    KJ_SWITCH_ONEOF(lazyDir) {
+      KJ_CASE_ONEOF(dir, kj::Own<jsg::fs::Directory>) {
+        return *dir;
+      }
+      KJ_CASE_ONEOF(func, kj::Function<kj::Own<jsg::fs::Directory>()>) {
+        auto dir = func();
+        lazyDir = kj::mv(dir);
+        return *lazyDir.get<kj::Own<jsg::fs::Directory>>();
+      }
+    }
+    KJ_UNREACHABLE;
+  }
+};
+}  // namespace
+
+kj::Own<jsg::fs::VirtualFileSystem> getWorkerFileSystem(
+    const FsMap& fsMap, kj::Own<jsg::fs::Directory> bundleDirectory) {
+  // Our root directory is a read-only directory
+  jsg::fs::Directory::Builder builder;
+  builder.addPath(fsMap.getBundlePath(), kj::mv(bundleDirectory));
+  builder.addPath(fsMap.getTempPath(), getTempDirectoryImpl());
+  return jsg::fs::newVirtualFileSystem(builder.finish());
+}
+
+kj::Own<jsg::fs::Directory> getTempDirectoryImpl() {
```

Comment:
I think we should follow posix naming style and call this tmp rather temp.

```suggestion
kj::Own<jsg::fs::Directory> getTmpDirectoryImpl() {
```

Reply:
I have a mild preference for `Temp` but don't feel strongly enough about it to argue :-)